### PR TITLE
Add search to website

### PR DESCRIPTION
**Pull request by aliabd**
Search! 

closes: #2121 

While this works really well as it is currently, I will probably need to do another pass to clean out all the regex. The challenge was that for docs especially some of the content exists in jsons and some in the templates. So to get the content in a searchable dictionary i need to render it first, which means i need to remove the html code using regex. 

There must be a better way to do this, and hopefully its a way that will let us link to specific ids on the page so that the user can get closer to the query. I'm open to changing a lot of the code, especially all the ugly stuff in the search-api, so please leave feedback! 

It searches through all the docs and guides

### CODE REVIEW

**Review by freddyaboulton — APPROVED** [review]
Just looked at the functionality and I think it works well @aliabd ! Sometimes the ranking of the matches was a bit off. For example, I'd expect the Oauth guide to be near the top when searching "oauth" but it's the sixth result. Something we can fix in v2 if it isn't an obvious fix.

image

**Review by abidlabs — APPROVED** [review]
Heck yeah @aliabd! Works super well in my tests

The only nit I have is: would it be possible to automatically select the first search result, so that you could press enter and it would navigate to that page without having to navigate down and select it?

**Comment by abidlabs** [comment]
I thought I commented about this, but it seems to have disappeared: navigating to the right section in the page. For some searches, like "iframe" or "WaveformOptions", it would be great if it navigated to the right section in the page. Overall, not a huge pain point though.

**Comment by abidlabs** [comment]
Ok so one thing I noticed is that when multiple items on a page match a given query, it only selects one of them to show in the search preview. That's fine, but we should prioritize headings over other items. For example, when I search "waveformoptions", I expect it will preview this part of the docs:

image

However, it shows the parameter instead:

image

and a user may completely miss the fact that there's a section dedicated to `gr.WaveformOptions`

File: js/_website/src/lib/components/search/search.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	return results
+		.map((index) => pages[index as number])
+		.map(({ slug, title, content, type }) => {
+			return {
+				slug,
+				title: replace_text_with_marker(title, match),
+				content: get_matches(content, match),
+				type
+			};
+		});
```
[the remark this answers is not in the document]
Comment (aliabd) [reply js/_website/src/lib/components/search/search.ts:42]: weird i get an error with your change: 

```
Argument of type '(index: number) => Page' is not assignable to parameter of type '(value: Id, index: number, array: Id[]) => Page'.
  Types of parameters 'index' and 'value' are incompatible.
    Type 'Id' is not assignable to type 'number'.
      Type 'string' is not assignable to type 'number'.ts(2345)
```

[the remark this answers is not in the document]
Comment (aliabd) [reply js/_website/src/lib/components/search/search.ts:42]: but i don't see an error otherwise

**Comment by aliabd** [comment]
@freddyaboulton @abidlabs you're right that the ranking is a bit off. Maybe on v2 I can rank by # of occurrences on the page. I don't think prioritizing headers makes sense for how the code is structured right now, but will look into it. Navigating to the right spot would be killer imo but i definitely will need to rework how i'm generating the content database for that to work. Will do that in v2. 

@abidlabs We can't autofocus the first result because it will take the cursor off the search box. 

@hannahblair thanks for the changes! addressed all except for the one on search.ts

**Comment by abidlabs** [comment]
> @abidlabs We can't autofocus the first result because it will take the cursor off the search box.

Not a big deal at all, but could pressing enter automatically go the first link?

**Comment by aliabd** [comment]
@abidlabs ok I made it work. First link is always in "focus" and pressing enter will go there, otherwise arrow down will move to the next link. Try it out and lmk what you think. I think it's pretty neat!

**Review by abidlabs — APPROVED** [review]
LGTM looks great to me @aliabd!

**Pull request merged**